File: src/components/ChatMessage.js
import { StyleSheet, View, Text, FlatList, SafeAreaView, StatusBar, Image } from "react-native";
import { useMemo } from "react";
import colors from "../constants/colors";
import { PROFILE_IMAGE_SIZE } from "../constants/sizes"
import Avatar from "./Avatar"

const DAY = 1000 * 3600 * 24;

const ChatMessage = ({ author, avatar, message, timestamp }) => {
    const timeText = useMemo(() => {
        if(timestamp === undefined) {
            return "???";
        }

        const now = new Date(); // TODO
        const then = new Date(timestamp);

        const getHourString = () => {
            const v = String(then.getHours() % 12);
            return (v == '0') ? '12' : v;
        }
        
        const getMinuteString = () => {
            return String(then.getMinutes()).padStart(2, '0');
        }

        const getTimeString = () => getHourString() + ":" + getMinuteString() + " " + ((then.getHours() >= 12) ? "PM" : "AM");
        const getDateString = () => then.toISOString().split("T")[0];

        then.toTimeString
        if( ((now - then) < DAY) && (now.getDate() == then.getDate()) ) {
            return "Today at " + getTimeString();
        }else if ( ((now - then) < 2*DAY) && (now.getDate() == (then.getDate()+1) ) ) {
            return "Yesterday at " + getTimeString();
        }else {
            return getDateString() + " " + getTimeString();
        }
    }, [timestamp]); // TODO: Add current day as dependency, so Today -> Yesterday at midnight

    return (
        <View style={chatStyles.chatMessageRoot}>
            <Avatar author={author} avatar={avatar} />
            <View style={chatStyles.messageR}>
                <View style={chatStyles.authorNameAndTime}>
                    <Text style={chatStyles.authorName}>{author}</Text>
                    <Text style={chatStyles.timeSent}>{timeText}</Text>
                </View>

                <Text style={chatStyles.messageText}>{message}</Text>
            </View>
        </View>
    )
};

const chatStyles = StyleSheet.create({
    chatMessageRoot: {
        padding: 8,
        width: "100%",
        maxWidth: "100%",
        flexDirection: "row",
    },

    messageR: {
        marginRight: PROFILE_IMAGE_SIZE+8,
        marginLeft: 8
    },

    authorNameAndTime: {
        flexDirection: "row",
        paddingBottom: 2
    },

    authorName: {
        color: colors.brandPrimary,
        fontWeight: "700",
        fontSize: 16,
        marginRight: 6
    },

    messageText: {
        color: colors.white,
        fontSize: 14,
    },

    timeSent: {
        color: colors.greyInactive,
        fontSize: 11,
        textAlignVertical: "bottom",
        fontWeight: "400",
        marginBottom: 2
    }
});


export default ChatMessage;

File: src/components/chat_message.rs
use iced::{Element, Text};
use iced::widget::{Column, Container, Row};
use iced::alignment::{Horizontal, Vertical};

use crate::components::avatar::Avatar;
use crate::constants::colors;
use crate::constants::sizes::PROFILE_IMAGE_SIZE;

pub struct ChatMessage {
    pub author: String,
    pub avatar: Option<String>,
    pub message: String,
    pub timestamp: Option<i64>,
}

impl ChatMessage {
    pub fn new(author: String, avatar: Option<String>, message: String, timestamp: Option<i64>) -> Self {
        ChatMessage { author, avatar, message, timestamp }
    }

    pub fn view(&self) -> Element<()> {
        let time_text = self.format_timestamp();

        Row::new()
            .push(Avatar::new(self.author.clone(), self.avatar.clone(), PROFILE_IMAGE_SIZE).view())
            .push(
                Column::new()
                    .push(
                        Row::new()
                            .push(Text::new(&self.author).size(16).color(colors::BRAND_PRIMARY))
                            .push(Text::new(&time_text).size(11).color(colors::GREY_INACTIVE))
                    )
                    .push(Text::new(&self.message).size(14).color(colors::WHITE))
            )
            .into()
    }

    fn format_timestamp(&self) -> String {
        if let Some(timestamp) = self.timestamp {
            let now = chrono::Utc::now().timestamp();
            let then = chrono::NaiveDateTime::from_timestamp(timestamp, 0);
            let duration = now - timestamp;

            if duration < 86400 && chrono::Utc::now().date() == then.date() {
                format!("Today at {}", then.format("%I:%M %p"))
            } else if duration < 172800 && chrono::Utc::now().date().pred() == then.date() {
                format!("Yesterday at {}", then.format("%I:%M %p"))
            } else {
                then.format("%Y-%m-%d %I:%M %p").to_string()
            }
        } else {
            "???".to_string()
        }
    }
}
